### ISSUE TITLE
Should ZIO#fromEither be By Name?

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2295]
Currently `ZIO.fromEither` takes its parameter by name and wraps it in `effectTotal`. This is to a certain extent safer but I would have expected `fromEither` to just life a pure `Either` value into a `ZIO`. This approach would also be more compositional because I could always wrap the `fromEither` call in `effectSuspendTotal` if I want the lazy behavior. Should we change this? The same issues also applies to other constructors such as `fromOption` and `fromTry`.

**Comment by neko-kai** [comment]
Previous discussion (when it was changed from strict to by-name): https://github.com/zio/zio/pull/577
^Note: I think the above was alone before tracing, there's an added aspect now that a lambda wrapper brings with it a line number

**Comment by jdegoes** [comment]
The laziness buys us the ability to "push" the errors into the ZIO runtime system.

Otherwise users write code like this:

```scala
def foo = {
  val zio = ZIO.fromEither(mySideEffectingMethodCall)
  ...
}
```

and now `foo` blows up, rather than `foo` returning a `ZIO`. 

ZIO has good mechanisms for error reporting and recovery so I think it makes sense to push as many failures as possible into ZIO.

**Comment by reibitto** [comment]
Just to add on to this, should the same principle apply for things like `someOrFail(...)` and so on? I'm not exactly sure where the line should be drawn and what would be deemed "consistent" when it comes to all these methods. Right now I don't have an intuition for which ones are/should be call-by-name and which shouldn't.

I actually ran into `someOrFail` that had error logging with untracked side-effects contained in it yesterday (still in the middle of a conversion to ZIO at work). I scratched my head for a while wondering how my `Some` was "failing" (it wasn't). That was totally a user error, but it did make me wonder about this issue.
